control-plane/pkg/reconciler/broker/broker.go

```diff
@@ -210,6 +213,17 @@ func (r *Reconciler) reconcileKind(ctx context.Context, broker *eventing.Broker)
 		}
 	}
 
+	// To make sure that our receivers have got at least the first broker configuration, we probe our receivers pods
+	ingress := brokerResource.Ingress.IngressType.(*contract.Ingress_Path)
+	if err := r.ProbeReceivers(ingress.Path); err != nil {
+		statusConditionManager.ProbeFailed(err)
+		// When a probe fails, we don't return an error but we requeue our broker so that we can probe it again later.
+		r.EnqueueAfter(broker, time.Duration(r.Configs.ProbeFailureRequeueDelayMs)*time.Millisecond)
+		return nil
+	} else {
+		statusConditionManager.ProbeSucceeded()
+	}
+
```

Comment:
I think this is problematic, because it's gonna perform a request to every receiver on every broker reconciliation. I think we should instead apply the same logic we use in eventing-kafka and in networking for performing probing, which reduces that load and applies that periodically

@devguyio is more expert than me on this :smile:

Reply:
As I said in a comment somewhere above, I plan to do a follow-up with a caching layer to cache the status.

Reply:
That's not really an optimization, but instead a different way to reconcile... Is it reasonable to do it here?

Reply:
Yeah, I understand but the end goal should be the same, it's just (IMO) simpler to reason about in the reconcile flow, the first time even if your just setting up your "out of flow thread" we have to do the request ASAP to set the status, so a synchronous probe makes sense to me here.

Once you have done the first request, we cache that pod p1 has responded for broker b1 with ok, so we don't probe it again for x time.

The "out of the flow thread" would periodically probe broker b1 (even if nothing changes in the system), we instead probe broker b1 once something changes in the system.

Reply:
> The "out of the flow thread" would periodically probe broker b1 (even if nothing changes in the system), we instead probe broker b1 once something changes in the system that "touches" that particular broker b1.

And that's my problem with this solution, if you have 10 brokers to reconcile one after another, this solution is going to be very expensive, because it's gonna ping all the receiver pods on every reconciliation loop. It will also slow down the reconciliation, maybe even considerably after a certain amount of receiver instances

Caching is also hard, because you need to keep that cache in sync, and there's no easy way to do it. While with probing, you just sync the cache periodically.

> it's just (IMO) simpler to reason about in the reconcile flow,

That might be true, on the other hand, i would love to reuse as much as possible the same pattern we have in eventing-kafka and networking.

This approach might also help us with the "future" problem of dispatcher readyness, similar to the one tackled in eventing-kafka

Reply:
Sorry, but do you mean that we should probe a receiver pod and assume that all brokers in that pod are good? 
That's not true, see linked issue's comments.

Regarding cache, what do you mean by in sync? In sync with what?

Reply:
Regarding consumer readiness eventing-kafka has 2 solution, 1 is probing (channel) and the other one (mtsource) is through kafka which is imo more reliable.

Reply:
Ok, sorry, I missed this. 

I meant that we probe broker b1 when something changes and the time x expires.

Reply:
Now, it should be clear that it's equivalent